### PR TITLE
Add API and algorithm to expose transform to workers

### PR DESCRIPTION
**Pull request by youennf**
***
<a href="https://pr-preview.s3.amazonaws.com/youennf/webrtc-insertable-streams/pull/62.html" title="Last updated on Feb 26, 2021, 5:01 PM UTC (146a751)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-insertable-streams/62/bc9ae2d...youennf:146a751.html" title="Last updated on Feb 26, 2021, 5:01 PM UTC (146a751)">Diff</a>

### CODE REVIEW

**Comment by youennf** [comment]
Fixes https://github.com/w3c/webrtc-insertable-streams/issues/18

**Review by alvestrand — CHANGES_REQUESTED** [review]
I'll have to flag this one as -1 until I figure out what it takes away.

**Comment by alvestrand** [comment]
Suggest making one PR to add the new API and another one that takes the old API away, so that we can discuss them separately.

**Comment by youennf** [comment]
> uggest making one PR to add the new API and another one that takes the old API away, so that we can discuss them separately.

I reverted the changes, PR now only adds stuff!

File: index.bs [hunk 1]
```diff
@@ -275,20 +227,112 @@ The <dfn method for="SFrameTransform">setEncryptionKey(|key|, |keyID|)</dfn> met
...
+
 The <dfn constructor for="RTCRtpScriptTransform" lt="RTCRtpScriptTransform(worker, options)"><code>new RTCRtpScriptTransform(<var>worker</var>, <var>options</var>)</code></dfn> constructor steps are:
 1. Set |t1| to an [=identity transform stream=].
 2. Set |t2| to an [=identity transform stream=].
 3. Set |this|.`[[writable]]` to |t1|.`[[writable]]`.
 4. Set |this|.`[[readable]]` to |t2|.`[[readable]]`.
-5. FIXME: transfer |t1|.`[[readable]]` and |t2|.`[[writable]]` to the dedicated worker.
-6. FIXME: Create counterpart of |this| in dedicated worker, for instance expose transfered |t1|.`[[readable]]` and |t2|.`[[writable]]`.
+5. Let |serializedOptions| be the result of [$StructuredSerialize$](|object|).
+6. Let |serializedReadable| be the result of [$StructuredSerializeWithTransfer$](|t1|.`[[readable]]`, « |t1|.`[[readable]]` »).
```
Comment (alvestrand) [inline index.bs:306]: Should the << marks be square brackets? (I think the second argument to Serialize is a transfer list)

**Review by alvestrand — APPROVED** [review]
Now that "removing the old way" is removed, this is a completion of the algorithm already landed, and I approve.

I do wonder if we can get the ability to "polyfill" the old behavior by allowing the "worker" argument to the ScriptTransform constructor to be the Javascript equivalent of "this" - that would make same-process processing possible, but not simpler than worker-based processing.

**Comment by alvestrand** [comment]
Note: I'd be even happier if the PR also added examples to explainer.md.

File: index.bs (hunk 1 above)
Comment (youennf) [reply index.bs:306]: Not sure what style is best here.
I reused the style used in https://streams.spec.whatwg.org/#ws-transfer.

**Comment by youennf** [comment]
> Note: I'd be even happier if the PR also added examples to explainer.md.

I agree we should do that before landing https://github.com/w3c/webrtc-insertable-streams/pull/64.

**Comment by alvestrand** [comment]
Thinking some more - in all cases the transform object is created on the main thread (constructor has exposed=window).
If "worker" argument is nullable, we can say "no event is fired when worker is null", and we have a very reasonably shimmable interface that gives us the old functionality. WDYT?

**Comment by alvestrand** [comment]
Note: DedicatedWorker is still a link that goes nowhere.

**Comment by guidou** [comment]
Some examples of how the API is intended to be used in practice would be very beneficial to be able to figure out a plan to implement it. I would prefer that we keep the old API in the spec until implementors have had time to evaluate the new one in practical settings.

File: index.bs [hunk 2]
```diff
@@ -275,20 +227,112 @@ The <dfn method for="SFrameTransform">setEncryptionKey(|key|, |keyID|)</dfn> met
...
+    readonly attribute RTCRtpScriptTransformer transformer;
+};
+
+[Global=(Worker,DedicatedWorker),Exposed=DedicatedWorker]
+partial interface DedicatedWorkerGlobalScope : WorkerGlobalScope {
+    attribute EventHandler onrtctransform;
+};
+
+[Global=(Worker,DedicatedWorker),Exposed=DedicatedWorker]
+interface RTCRtpScriptTransformer {
```
Comment (alvestrand) [inline index.bs:285]: After having tried to comprehend the overall algorithm several times, I've concluded that we should rename this object to RTCRtpScriptTransformSocket. It's not a transformer itself, it's what you plug a transform (like SframeTransform) into.

File: index.bs [hunk 3]
```diff
@@ -275,20 +227,112 @@ The <dfn method for="SFrameTransform">setEncryptionKey(|key|, |keyID|)</dfn> met
...
+    readonly attribute ReadableStream readable;
+    readonly attribute WritableStream writable;
+    readonly attribute any options;
+};
+
 [Exposed=(Window)]
 interface RTCRtpScriptTransform {
```
Comment (alvestrand) [inline index.bs:292]: similarly, RTCRtpScriptTransform is not a transform. I suggest we call it RTCRtpScriptTransformController.

**Comment by alvestrand** [comment]
New issues:
- Add examples
- Work on naming (align with stream spec, such as DuplexStream and Endpoint Pair)
-

File: index.bs [hunk 4]
```diff
@@ -275,20 +227,112 @@ The <dfn method for="SFrameTransform">setEncryptionKey(|key|, |keyID|)</dfn> met
...
+
+// New interfaces to expose JavaScript-based transforms.
+
+[Global=(Worker,DedicatedWorker),Exposed=DedicatedWorker]
+interface RTCTransformEvent : Event {
+    readonly attribute RTCRtpScriptTransformer transformer;
+};
+
+[Global=(Worker,DedicatedWorker),Exposed=DedicatedWorker]
+partial interface DedicatedWorkerGlobalScope : WorkerGlobalScope {
```
Comment (foolip) [inline index.bs:280]: With partial interfaces in Web IDL the parent interface shouldn't be specified. I've sent https://github.com/w3c/webrtc-insertable-streams/pull/71 to fix this.

Comment (foolip) [reply index.bs:280]: Fixed in https://github.com/w3c/webrtc-insertable-streams/pull/70.

File: index.bs [hunk 5]
```diff
@@ -62,54 +62,6 @@ an additional API on {{RTCRtpSender}} and {{RTCRtpReceiver}} to
 insert the processing into the pipeline.
 
 <pre class="idl">
-// New dictionary.
-dictionary RTCInsertableStreams {
```
Comment (foolip) [inline index.bs:66]: The `RTCInsertableStreams` dictionary was removed here, but it's still used as the return type of two methods in https://w3c.github.io/webrtc-insertable-streams/#specification.

Comment (youennf) [reply index.bs:66]: Oh, that is a mistake from splitting  the PR, I will add it back.

Comment (youennf) [reply index.bs:66]: https://github.com/w3c/webrtc-insertable-streams/pull/73

**Pull request merged**